src/panels/config/integrations/integration-panels/bluetooth/bluetooth-device-page.ts

```diff
@@ -0,0 +1,183 @@
+import type { CSSResultGroup, PropertyValues, TemplateResult } from "lit";
+import { html, LitElement } from "lit";
+import { customElement, property, state } from "lit/decorators";
+import memoizeOne from "memoize-one";
+import type { UnsubscribeFunc } from "home-assistant-js-websocket";
+import type { HASSDomEvent } from "../../../../../common/dom/fire_event";
+import type { LocalizeFunc } from "../../../../../common/translations/localize";
+import type {
+  DataTableColumnContainer,
+  DataTableRowData,
+  RowClickedEvent,
+} from "../../../../../components/data-table/ha-data-table";
+import "../../../../../components/ha-fab";
+import "../../../../../components/ha-icon-button";
+import "../../../../../layouts/hass-tabs-subpage-data-table";
+import { haStyle } from "../../../../../resources/styles";
+import type { HomeAssistant, Route } from "../../../../../types";
+
+interface DeviceRowData extends DataTableRowData {
+  address: string;
+  connectable: boolean;
+  manufacturer_data: { [id: string]: any };
+  name: string;
+  rssi: number;
+  service_data: { [id: string]: any };
+  service_uuids: string[];
+  source: string;
+  time: number;
+  tx_power: number;
+}
+
+interface RemoveDeviceRowData {
+  address: string;
+}
+
+interface EventDataMessage {
+  add?: DeviceRowData[];
+  change?: DeviceRowData[];
+  remove?: RemoveDeviceRowData[];
+}
+
+const subscribeBluetoothAdvertisements = (
+  hass: HomeAssistant,
+  callback: (event: EventDataMessage) => void
+): Promise<UnsubscribeFunc> =>
+  hass.connection.subscribeMessage(callback, {
+    type: `bluetooth/subscribe_advertisements`,
+  });
+
+@customElement("bluetooth-device-page")
+export class BluetoothDevicePage extends LitElement {
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @property({ attribute: false }) public route!: Route;
+
+  @property({ type: Boolean }) public narrow = false;
+
+  @property({ attribute: "is-wide", type: Boolean }) public isWide = false;
+
+  @state() private _rows: { [id: string]: DeviceRowData } = {};
+
+  private _unsub?: Promise<UnsubscribeFunc>;
+
+  private _firstUpdatedCalled = false;
+
+  public connectedCallback(): void {
+    super.connectedCallback();
+    if (this.hass && this._firstUpdatedCalled) {
+      this._unsub = subscribeBluetoothAdvertisements(
+        this.hass,
+        this._handleIncomingEventDataMessage
+      );
+    }
+  }
+
+  protected firstUpdated(changedProperties: PropertyValues): void {
+    super.firstUpdated(changedProperties);
+    if (this.hass) {
+      this._unsub = subscribeBluetoothAdvertisements(
+        this.hass,
+        this._handleIncomingEventDataMessage
```

Comment:
```suggestion
        this._handleIncomingEventDataMessage.bind(this)
```

I had to add a bind as otherwise `this._rows` was undef